### ISSUE TITLE
Odd and undesired behavior when adding a semi-transparent background layer above other layers

### ISSUE THREAD

**Issue by danielamitay** [mapbox/mapbox-gl-native#5832]
**Platform:** iOS
**Mapbox SDK version:** ios-v3.3.2 

### Steps to trigger behavior

 1. Take any map style (e.g. Streets)
 2. Add a new background layer to be the topmost layer
 3. Make it's opacity anything between 0 and 1 (e.g. 0.5)

Simple example, only one layer added:
![screen shot 2016-07-29 at 1 02 57 pm](https://cloud.githubusercontent.com/assets/647626/17256796/10ebd8ba-558d-11e6-8614-5e79d58336a0.png)

### Expected behavior

![simulator screen shot jul 29 2016 2 44 58 pm](https://cloud.githubusercontent.com/assets/647626/17259795/270fbfe0-559b-11e6-9468-31b066da9882.png)

### Actual behavior
(side note: Studio/gl-js has a different behavior than gl-native as well)

![simulator screen shot jul 29 2016 1 07 41 pm](https://cloud.githubusercontent.com/assets/647626/17259788/203a1f62-559b-11e6-8551-b37fdf2ae080.png)

There are additional rendering issues at certain zoom levels:

From 11 to 11.5, and from 15.5 to 16.5:
![11-11 5_250](https://cloud.githubusercontent.com/assets/647626/17260101/85e51b22-559c-11e6-8690-d87b7a934627.gif) ![15 5-16 5_250](https://cloud.githubusercontent.com/assets/647626/17260104/88b41934-559c-11e6-97dc-8fc81a9f356b.gif)

**Comment by danielamitay** [comment]
Some more info:

Here is what it looks like in Studio:

![screen shot 2016-07-29 at 4 17 05 pm](https://cloud.githubusercontent.com/assets/647626/17262216/1f898eb0-55a8-11e6-9652-9a520a2666bb.png)

I tried setting the color to have an alpha channel instead of the layer, and the Studio rendered it correctly (however, rendering a semi-transparent background layer should still be supported):
![screen shot 2016-07-29 at 4 17 37 pm](https://cloud.githubusercontent.com/assets/647626/17262231/31d476e8-55a8-11e6-98df-3eb8cca5c9c3.png)

Unfortunately, doing a background layer with opacity 1 and fill color with an alpha channel looks like this on the device (and still has the visual glitches from the gifs above):

![simulator screen shot jul 29 2016 4 17 40 pm](https://cloud.githubusercontent.com/assets/647626/17262303/9320c514-55a8-11e6-9714-80d3f9c19746.png)

**Comment by tmpsantos** [comment]
Trying to repro here. On the editor it is exactly how you described. Same goes for android `android-v4.1.1`.

Good news is it works just fine on native with the tip of the tree (472ab7442f5).

![screenshot_20160801-182731_1024](https://cloud.githubusercontent.com/assets/996307/17299167/da933606-5815-11e6-9054-b2a53fc9cf4b.png)
![screenshot_from_2016-08-01_18-04-44](https://cloud.githubusercontent.com/assets/996307/17299168/da97662c-5815-11e6-8bb7-d806095efc87.png)

**Comment by tmpsantos** [comment]
This issue was a problem with a shader.

```
[ERROR] {mapbox-glfw}[OpenGL]: GL_DEBUG_SOURCE_API GL_DEBUG_TYPE_ERROR 54 GL_DEBUG_SEVERITY_HIGH - GL_INVALID_OPERATION in glUniform4("u_ratio"@1 has 1 components, not 4)
terminate called after throwing an instance of 'mbgl::gl::Error'
  what():  glUniform4fv(location, 1, t.data()): Error GL_INVALID_OPERATION - ../../src/mbgl/shader/uniform.cpp:27
```

Closing this ticket since the bug was fixed on the master branch.

**Comment by tmpsantos** [comment]
Create a ticket on our test suite to add a test case like this, so I can also verify if this got fixed on mapbox-gl-js too. I would say probably yes since we share the same shaders.

**Comment by tmpsantos** [comment]
For the record, it was fixed by a17b0bcd82ebe86.
